lib/src/utils.dart

```diff
@@ -82,3 +83,73 @@ void logDependencyInfractions(String infraction, Iterable<String> dependencies)
   final sortedDependencies = dependencies.toList()..sort();
   logger.warning([infraction, bulletItems(sortedDependencies), ''].join('\n'));
 }
+
+/// Lists the packages with infractions
+List<String> getDependenciesWithPins(Map dependencies) {
+  final List<String> infractions = [];
+  for (String packageName in dependencies.keys) {
+    String version;
+    if (dependencies[packageName] is String) {
+      version = dependencies[packageName];
+    } else {
+      final Map packageMeta = dependencies[packageName];
+      if (packageMeta.containsKey('version')) {
+        version = packageMeta['version'];
+      } else {
+        // This feature only works for versions, not git refs or paths.
+        break;
+      }
+    }
+
+    if (doesVersionPinDependency(version)) {
+      infractions.add(packageName);
+    }
+  }
+
+  return infractions;
+}
+
+/// Returns whether the version restricts patch or minor upgrades.
+bool doesVersionPinDependency(String rawVersion) {
+  final String version = rawVersion.replaceAll('"', '').replaceAll("'", '');
+
+  final VersionConstraint constraint = new VersionConstraint.parse(version);
+
+  // No max set = no pin
+  if (constraint.isAny) {
+    return false;
+  }
+
+  // Direct pin
+  if (constraint is Version) {
+    return true;
+  }
+
+  if (constraint is VersionRange) {
+    // Defined maximum bound is a pin
```

Comment:
Just thought... it may not be immediately obvious to consumers why this is considered a pin. 

It may be helpful to make the failure message be more verbose, and outline all the cases that are considered pins and why.